File: api/status.js
const crypto = require("crypto");
const https = require("https");

function base64UrlEncode(str) {
  return Buffer.from(str)
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=/g, "");
}

function createJWT(payload, privateKey, kid, aud) {
  const header = {
    alg: "ES256",
    typ: "JWT",
    kid: kid,
  };

  const iat = Math.floor(Date.now() / 1000);
  const fullPayload = {
    ...payload,
    iat: iat,
    exp: iat + 120, // Valid for 2 minutes
    jti: crypto.randomUUID(),
    aud: aud,
    iss: "_ELmUvm5LOKEBjp0-TLBe4_J8iC9J0lQ",
    sub: "_ELmUvm5LOKEBjp0-TLBe4_J8iC9J0lQ",
  };

  const encodedHeader = base64UrlEncode(JSON.stringify(header));
  const encodedPayload = base64UrlEncode(JSON.stringify(fullPayload));
  const signatureInput = `${encodedHeader}.${encodedPayload}`;

  const signer = crypto.createSign("SHA256");
  signer.update(signatureInput);
  const signature = signer
    .sign({ key: privateKey, dsaEncoding: "ieee-p1363" }, "base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=/g, "");

  return `${signatureInput}.${signature}`;
}

module.exports = async (req, res) => {
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "GET, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type");

  if (req.method === "OPTIONS") {
    return res.status(200).end();
  }

  const { id, exchange_code } = req.query;
  if (!id || !exchange_code) {
    return res.status(400).json({ error: "Missing sign_request_id or exchange_code" });
  }

  const clientId = "_ELmUvm5LOKEBjp0-TLBe4_J8iC9J0lQ";
  const kid = "key-1";
  const privateKey = `-----BEGIN PRIVATE KEY-----
MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgNbVMxiHb2ODp6/Yw
CmSfkYQoenLG7keDDINXGtTOGR6hRANCAASfQOloP4YWjS+pF5aWVsshFXahP4j9
bQotHZrdaiEpoWTtcaE/jxqjhU8t0pY6Yy7PFGY7l0jCFTOwtIj6pC50
-----END PRIVATE KEY-----`;

  const apiUrl = `https://app.sign.singpass.gov.sg/api/v3/sign-requests/${id}/signed-doc`;

  const jwt = createJWT(
    {
      client_id: clientId,
      exchange_code: exchange_code
    },
    privateKey,
    kid,
    apiUrl
  );

  const options = {
    method: "GET",
    headers: {
      Authorization: jwt,
      Accept: "application/json",
    },
  };

  const apiReq = https.request(apiUrl, options, (apiRes) => {
    let data = "";
    apiRes.on("data", (chunk) => (data += chunk));
    apiRes.on("end", () => {
      try {
        if (
          apiRes.headers["content-type"] &&
          apiRes.headers["content-type"].includes("application/json")
        ) {
          const result = JSON.parse(data);
          if (apiRes.statusCode === 200) {
            res.status(200).json({ status: "signed", signed_doc_url: result.signed_doc_url });
          } else if (apiRes.statusCode === 400 && result.error === "DOCUMENT_NOT_SIGNED") {
            res.status(200).json({ status: "pending" });
          } else if (apiRes.statusCode === 400 && result.error === "REQUEST_EXPIRED") {
            res.status(200).json({ status: "expired", error: "Signing request has expired" });
          } else if (apiRes.statusCode === 400 && result.error === "REQUEST_CANCELLED") {
            res.status(200).json({ status: "cancelled", error: "Signing request was cancelled" });
          } else if (apiRes.statusCode === 400 && result.error === "REQUEST_FAILED") {
            res.status(200).json({ status: "failed", error: "Signing request failed" });
          } else {
            res.status(200).json({ status: "failed", error: result.error || "Unknown error", details: result });
          }
        } else {
          res
            .status(apiRes.statusCode)
            .json({ error: "API returned non-JSON response", raw: data });
        }
      } catch (e) {
        res.status(500).json({ error: "Failed to parse API response", raw: data });
      }
    });
  });

  apiReq.on("error", (e) => {
    res.status(500).json({ error: e.message });
  });

  apiReq.end();
};

File: api/status.py
"""
api/status.py — GET /api/status?id=<sign_request_id>
Polls Singpass for signing result.
"""
import json, ssl, urllib.parse
import http.client
from http.server import BaseHTTPRequestHandler
from api._shared import (
    get_api_base, build_auth_token, json_response,
    get_env, SIGN_ENDPOINT, _store
)


class Handler(BaseHTTPRequestHandler):

    def log_message(self, format, *args): pass

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()

    def do_GET(self):
        params     = dict(urllib.parse.parse_qsl(urllib.parse.urlparse(self.path).query))
        req_id     = params.get("id", "")
        is_staging = params.get("staging", "1") == "1"

        record = _store.get(req_id)
        if not record:
            json_response(self, 404, {"error": "Session not found"})
            return

        if record["status"] == "signed":
            json_response(self, 200, {"status": "signed", "signed_doc_url": record["signed_doc_url"]})
            return

        client_id, private_key_pem, kid, _ = get_env()
        api_base   = get_api_base(is_staging)
        result_url = f"{api_base}{SIGN_ENDPOINT}/{req_id}/result"

        try:
            auth_token    = build_auth_token(client_id, private_key_pem, kid, result_url)
            exchange_code = record["exchange_code"]
            body_data     = json.dumps({"exchange_code": exchange_code}).encode()

            parsed = urllib.parse.urlparse(result_url)
            ctx    = ssl.create_default_context()
            conn   = http.client.HTTPSConnection(parsed.netloc, timeout=15, context=ctx)
            try:
                conn.request("POST", parsed.path, body=body_data, headers={
                    "Authorization": f"Bearer {auth_token}",
                    "Content-Type":  "application/json",
                    "Content-Length": str(len(body_data)),
                })
                resp   = conn.getresponse()
                result = json.loads(resp.read())
            finally:
                conn.close()

            signed_url = result.get("signed_doc_url") or result.get("download_url")
            if signed_url:
                _store[req_id].update({"status": "signed", "signed_doc_url": signed_url})
                json_response(self, 200, {"status": "signed", "signed_doc_url": signed_url})
            else:
                json_response(self, 200, {"status": "pending"})

        except (http.client.HTTPException, OSError):
            json_response(self, 200, {"status": "pending"})
        except Exception as e:
            json_response(self, 200, {"status": "pending", "debug": str(e)})